Add acts as follower

daisakuhazui opened the pull request:
# WHAT
acts_as_followerというgemを用いてUserにフォロー機能を実装。

# WHY
質問（Questionモデル）を投稿するWebアプリを作成していいますが、UserモデルがQuestionモデルの親要素であるため。

Commit:
followアクションの追加

Commit:
Followモデルの作成

acts_as_followerの導入に伴い、Followモデルを作成しました。

Commit:
Userモデルにフォロー機能を実装する

フォローする・フォローされる機能をUserモデルに追加しました。

Commit:
フォロー数とフォロワー数の表示パーシャル

プロフィール画面とタイムラインで、current_userのフォロー数・フォロワー数を表示させるためのパーシャルを作成しました。

Commit:
acts_as_followerをGemfileに追加

Commit:
Followモデルのマイグレーションファイル

Commit:
ユーザーのプロフィール画面にフォロー関連パーシャルの追加

ユーザーのプロフィール画面にフォロー数とフォロワー数が表示されるようにしました。

また、current_user以外にはFollowボタン、Unfollowボタンが表示されるようにする予定です。（18行目）

Commit:
フォローボタンのフォーム

current_user以外のプロフィール画面でボタンが表示されるようにする予定です。

if文を用い、ユーザーをフォローしている場合とフォローしていない場合で表示させるボタンを変化させます。

Commit:
Followボタン

ユーザーをフォローしていない場合、プロフィール画面に表示させるFollowボタンです。

Commit:
Unfollowボタン

ユーザーをフォローしている場合にプロフィール画面に表示させるボタンです。

ArakawaShintaro commented on app/views/shared/_stats.html.erb:
```diff
@@ -0,0 +1,5 @@
+<% @user ||= current_user %>
```
ここは変数にいれなくてもcurrent_userでいいんじゃないでしょうか

ArakawaShintaro commented on app/views/users/_follow.html.erb:
```diff
@@ -0,0 +1,6 @@
+<%= form_for(current_user.followers.build(follower_id: @user.id)) do |f| %>
```
ここのエラーとしては下記が、空の配列になっているためbuildできないというようなエラーですよね
```
current_user.followers
```

ArakawaShintaro commented on app/views/users/_follow_form.html.erb:
```diff
@@ -0,0 +1,9 @@
+<% unless current_user.id == @user %>
```
ここって、@user.idじゃないといけない気がします

ArakawaShintaro commented on app/views/users/show.html.erb:
```diff
@@ -14,3 +14,5 @@
 </div>
 <!-- ユーザーが投稿した質問もタイムライン形式で表示させる -->
+<%= render 'shared/stats' %>
+<%= render 'users/follow_form' if user_signed_in? %>
```
render先で@userを使っていると思うのですが、ここで@userを渡さなくても変数はきちんと参照できていますか？

Commit:
FollowsContorollerの作成

createアクションとdestroyアクションを定義しました。

Commit:
一部コードをViewからControllerへ移動

テンプレートファイルはあくまで結果の表示に徹するため、パーシャルで記述していた内容をアクション内に移動させました。
また、理解不足でとりあえず記述してたfollowアクションは削除しました。

Commit:
不要な改行の削除

Commit:
コードを結果の表示のみにしました

1行目はUsersContorollerのshowアクションに記述しました。

Commit:
フォローボタンの設置

@follow_userはUsersControllerのshowアクションで定義しています。
また、FollowsControllerにパラメータを渡すためhidden_field_tagを追加しました。

Commit:
Unfollowボタンの設置

@follow_userはUsersControllerのshowアクションで定義しています。
また、FollowsControllerにパラメータを渡すためhidden_field_tagを追加しました。

Commit:
自分のプロフィール画面でフォローボタンを非表示

自分のプロフィール画面でフォローボタンが出現していた箇所を修正しました。

Commit:
二重に記述されていた箇所を削除

unless以下のコードがかぶっていたので削除しました。

Commit:
フォロー機能関連のルーティングを追加

Commit:
gemのバージョンアップ

ArakawaShintaro commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,19 @@
+  before_action :set_follow
+
+  def create
+    current_user.follow(@follow)
```
ここもし、ログインしていない人がfollowしたらどうなるんでしょうか？
current_userがとれなくてエラーがでてしまいますね。
ログインしてなければfollowボタンは非表示にする感じですかね？

daisakuhazui replied:
ログインしていない場合、followボタンは非表示にします。

ArakawaShintaro commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def destroy
+    current_user.stop_following(@follow)
```
こちらも同上です

daisakuhazui replied:
同上です。非表示にします。

ArakawaShintaro commented on app/controllers/follows_controller.rb:
```diff
@@ -0,0 +1,19 @@
+
+  private
+
+  def set_follow
```
いいっすね 👍

daisakuhazui replied:
DRYを意識する余裕が出てきました✌️

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -2,6 +2,9 @@ class UsersController < ApplicationController
   before_action :set_user
 
   def show
+    @user ||= current_user
```
ここって変数に入れる意味ってありますか！？

daisakuhazui replied:
@userがnil、つまり自分のプロフィールを見た時にcurrent_userを自己代入すれば、フォロー数の表示がスムーズかなと思いました。（もちろん、実際はそんなことはないのですが）

でも今しがた試してみると、いちいち@userに代入しなくてもフォロー数取れていました・・・削除します。

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,6 +1,9 @@
 Rails.application.routes.draw do
+  get "follows/create"
```
こちら、resourcesでfollow使ってるのに必要ですか？
また、create destoryそれぞれがgetってのもおかしいですね。
それぞれpost deleteにしないといけません。

daisakuhazui replied:
おっしゃるように、不要です。
自動生成されたコードを放置しておりました。

ArakawaShintaro reviewed (commented):
ひとまず見ました！

Commit:
不要な変数の記述を削除

Commit:
ルーティングの調整

Commit:
サインインすればフォローボタンが表示される仕様に変更

Commit:
current_userが存在しない場合のエラーを修正

サインインしていない場合フォローボタンは表示されない仕様にしてあるため、インスタンス変数でレコードを渡す必要がないためif文で条件付けを行いました。
（UsersControllerのshowアクションが呼び出された際、サインインしていないとcurrent_userがnilになりエラーが起こったため）

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -2,8 +2,11 @@ class UsersController < ApplicationController
+    if signed_in?
+      @follow_user = current_user.follows.build(followable_id: @user.id, follower_id: :current_user)
+      @unfollow_user = current_user.follows.find_by(followable_id: @user.id)
+    else
```
elseの時は処理はどうしますか？

daisakuhazui replied:
確認ありがとうございます。

elseの時はフォローボタンを表示しないので、特に処理は行いません。
・・・ということは、elseの記述がいらんということでしょうかね。試したところ、elseを削除しても問題なく動きましたので。

ArakawaShintaro reviewed (commented):
1点だけ気になりました！

Commit:
elseを削除

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -5,7 +5,7 @@ def show
       @follow_user = current_user.follows.build(followable_id: @user.id, follower_id: :current_user)
       @unfollow_user = current_user.follows.find_by(followable_id: @user.id)
-    else
+
```
無駄な空行が...

Commit:
空行の削除

ArakawaShintaro reviewed (approved):
こう言う細かいところは気をつけましょう(>_<)
おっけーです！

daisakuhazui commented:
@ArakawaShintaro 
了解です。確認ありがとうございます！

Commit:
フォロー機能に関するアクションの追加

自分がフォローしているユーザー、自分をフォローしているユーザーを取得できるようにしました。
また、@titleで、遷移先のページにページタイトルを渡しています。

Commit:
フォロー数の数字にリンクを追加

ユーザーのフォロー数をクリックすると、フォロー一覧のページに飛べるようにしました。

Commit:
FollowingとFollowersページの作成

フォローしているユーザーの一覧を表示するページを作成しました。

Commit:
メンバールーティングの追加

リソースブロックに、show_followingとshow_followersを追加しました。

ArakawaShintaro reviewed (approved):
LGTM
